File: erlang/slides/src/haskell/factorial.hs
#!/usr/bin/env runhaskell

import Control.Monad.State
import Text.Printf

fact_rec :: Int -> State Int Int
fact_rec 0 = do
    acc <- get
    return acc
fact_rec n = do
    acc <- get
    put (acc * n)
    fact_rec (n - 1)

{-fact_rec :: Int -> State Int Int
fact_rec 0 =
    get >>= \acc ->
        return acc
fact_rec n = 
    get >>= \acc ->
        put (acc * n) >>
            fact_rec (n - 1)-}

fact :: Int -> Int
fact n = evalState (fact_rec n) 1

main = do
    putStrLn "Factorial:"
    printf "fact(6) = %d\n" (fact 6)

File: erlang/slides/src/erlang/factorial.erl
-module(factorial).

-compile({parse_transform, erlmonads_expr_do}).
-compile({no_auto_import,[get/0]}).

-export([
    fact/1
]).

-import(erlmonads_state, [
    get/0,
    put/1
]).


fact_rec(0) ->
    do([erlmonads_state ||
        Acc <- get(),
        return(Acc)
    ]);
fact_rec(N) ->
    do([erlmonads_state ||
        Acc <- get(),
        put(Acc * N),
        fact_rec(N - 1)
    ]).

fact(N) ->
    R = erlmonads_state:eval(
        fact_rec(N),
        1
    ),
    R.
